F1-Beta Score

SSaishruthi opened the pull request:
Scripts for F1 Beta Score.

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -66,26 +71,52 @@ class are returned.
         super(F1Score, self).__init__(name=name)
         self.num_classes = num_classes
+        if beta <= 0:
+            raise ValueError("beta value should be greater tha zero")
```
Typo: tha -> than

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -206,3 +238,97 @@ def reset_states(self):
+    """Computes F-Beta Score.
+
+    This is the weighted harmonic mean of precision and recall.
+    Output range is [0,1].
```
Nit: `[0,1]` -> `[0, 1]`

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -206,3 +238,97 @@ def reset_states(self):
+                 average,
+                 name='fbeta_score',
+                 dtype=tf.float32):
+        super().__init__(num_classes, beta, average, name=name)
```
Pass `dtype` to `super` constructor.

Squadrick commented on tensorflow_addons/metrics/BUILD:
```diff
@@ -54,3 +54,16 @@ py_test(
 )
+
+py_test(
+    name = "f1beta_test",
```
Rename: `f1beta_test` -> `fbeta_test`. 

Rename the file as well.

SSaishruthi commented:
@Squadrick 
Updated PR

WindQAQ commented:
hmm, I think the inheritance is somewhat not ideal. IMO, we should do like this
```python
# Expose beta argument for FBetaScore
class FBetaScore(Metric):
    def __init__(self,
                 num_classes,
                 beta,
                 average,
                 name='fbeta_score',
                 dtype=tf.float32):
    ...
    # implement update_state etc.

# Inherit from FBetaScore and pass beta=1.0
# Do not expose beta argument 
class F1Score(FBetaScore):
    def __init__(self,
                 num_classes,
                 average,
                 name='f1_score',
                 dtype=tf.float32):
        super(F1Score, self).__init__(num_classes, 1.0, average, name, dtype)
```
What do you think @Squadrick and @SSaishruthi ?

Squadrick commented:
@WindQAQ Yup, look at [this](https://github.com/scikit-learn/scikit-learn/blob/7813f7efb5b2012412888b69e73d76f2df2b50b6/sklearn/metrics/classification.py#L950) for reference.

@SSaishruthi

SSaishruthi commented:
Thanks @Squadrick @WindQAQ for the comments.

Following are my to do as per the suggestion:

- [ ] Update FBeta and F1 inheritance.
- [ ] Update test cases.
- [ ] Doc update for the same

I will keep you posted as I update scripts.

SSaishruthi commented:
Updated scripts as per the comments.

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -23,69 +23,107 @@
+        super(FBetaScore, self).__init__(name=name)
         self.num_classes = num_classes
+        # type check
+        if not isinstance(beta, float):
```
`beta` could be a scale tensor of dtype `tf.float`

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -165,10 +203,11 @@ def result(self):
         r_sum = tf.cast(self.true_positives + self.false_negatives, tf.float32)
```
Use `dtype`

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -206,3 +246,103 @@ def reset_states(self):
+    recall. Output range is [0, 1]. This works for both
+    multi-class and multi-label classification.
+
+    F-Beta = (2) * ((precision * recall) / (precision + recall))
```
Nit: `F-Beta` -> `F-1`

Squadrick commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -206,3 +246,103 @@ def reset_states(self):
+                 Acceptable values are `None`, `micro`, `macro` and
+                 `weighted`.
+                 Default value is None.
+       beta : `float`
```
Remove from docstring.

SSaishruthi commented:
@Squadrick 
Addressed comments. For now,
1. Added check for const tensor
2. Added a note in doc that it can also be a const tensor of ftype float32. 
Let me know if this is ok

WindQAQ commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -23,69 +23,109 @@
+            self.beta = beta
         if average not in (None, 'micro', 'macro', 'weighted'):
             raise ValueError("Unknown average type. Acceptable values "
                              "are: [micro, macro, weighted]")
```
in fact, we also accept None, right?

WindQAQ commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -206,3 +248,103 @@ def reset_states(self):
+
+
+class F1Score(FBetaScore):
+    """Calculates F1 micro, macro or weighted based on the user's choice.
```
I think 
Computes F-1 Score.
is necessary as you use this in F-beta.

WindQAQ commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -23,69 +23,109 @@
+         average : Type of averaging to be performed on data.
+                   Acceptable values are None, micro, macro and
+                   weighted.
+         beta : float
```
float. Determines the weight of precision ...

WindQAQ commented on tensorflow_addons/metrics/f1_scores.py:
```diff
@@ -206,3 +248,103 @@ def reset_states(self):
+
+    def __init__(self, num_classes, average, name='f1_score',
+                 dtype=tf.float32):
+        super().__init__(num_classes, average, 1.0, name=name, dtype=dtype)
```
lets make the style unified in the same file though this is also correct. Use
```python
super(F1Score, self).__init__(...)
```
instead.

SSaishruthi commented:
Thanks, @WindQAQ 

Addressed comments except the following. 

1. Combining test cases.
2. File name.

I am fine with changing file name but I am not sure if which will be ideal for test cases. 

@Squadrick

Squadrick commented:
@SSaishruthi Call it `f_scores.py`. 

Having the test cases in two separate files or a single file is your call. If you want to combine them, you can add the case to `FBetaTest` for `beta=1.0` rather than copying from one file to the other.

SSaishruthi commented:
Addressed comments. I am thinking of having test separately if that is ok. 

Thought of combining but I got confused thinking if it is ok not to test `FScore` directly but instead use `FBetaScore` though we inherit from FBeta. Would like to test without providing `Beta` value too. 
Considering all these, I did not modify the code. 

Can you please let me know if the changes are ok?

Squadrick commented on tensorflow_addons/metrics/fbeta_test.py:
```diff
@@ -0,0 +1,117 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Tests F beta metrics."""
```
Nit: `Tests F beta metrics` -> `Tests for F-Beta metric`.

WindQAQ commented on tensorflow_addons/metrics/f_scores.py:
```diff
@@ -23,72 +23,112 @@
+               beta=0.4, average=None)
+    fb_score.update_state(actuals, preds)
+    print('F1-Beta Score is: ',
+         fb_score.result().numpy()) # [1. 0. 0.]
```
Please lint the docstring above. See the docstring of `F1Score` for more details.

Squadrick replied:
@SSaishruthi Just need to address this.

SSaishruthi replied:
@Squadrick Thanks for pointing this issue. I misunderstood @WindQAQ comment to be only in this line. Updated now.

WindQAQ commented on tensorflow_addons/metrics/f_scores.py:
```diff
@@ -206,3 +248,104 @@ def reset_states(self):
+        base_config = super(F1Score, self).get_config()
+        new_config = dict(list(base_config.items()))
+        del new_config['beta']
+        return new_config
```
I think this could be done by:
```python
def get_config(self):
    base_config = super(F1Score, self).get_config()
    del base_config["beta"]
    return base_config
```

SSaishruthi commented:
Thanks, @WindQAQ 

@Squadrick 
Updated scripts. Can we trigger the tests now?

seanpmorgan commented:
> Thanks, @WindQAQ
> 
> @Squadrick
> Updated scripts. Can we trigger the tests now?

Sanity check is failing just FYI

SSaishruthi commented:
@seanpmorgan Thanks for posting
I have made changes to the scripts. Can we trigger the test again?

SSaishruthi commented:
Thanks @Squadrick 
Looks like tests have passed

SSaishruthi commented:
@Squadrick Updated scripts
